lib/auth/webauthncli/fido2.go

```diff
@@ -0,0 +1,557 @@
+var fidoDeviceLocations = libfido2.DeviceLocations
+var fidoNewDevice = func(path string) (FIDODevice, error) {
+	return libfido2.NewDevice(path)
+}
+
+// LoginPrompt is the user interface for FIDO2Login.
+type LoginPrompt interface {
+	// PromptPIN prompts the user for their PIN.
+	PromptPIN() (string, error)
+	// PromptAdditionalTouch prompts the user for an additional security key
+	// touch.
+	// Additional touches may be required after PINs and during passwordless flows.
+	PromptAdditionalTouch() error
+}
+
+// FIDO2Login signs an assertion using available CTAP1 or CTAP2 devices.
+// It must be called with a context with timeout, otherwise it can run
+// indefinitely.
+// The informed user is used to disambiguate credentials in case of passwordless
+// logins.
+// It returns an MFAAuthenticateResponse and the credential user, if a resident
+// credential is used.
+func FIDO2Login(
+	ctx context.Context,
+	origin, user string, assertion *wanlib.CredentialAssertion, prompt LoginPrompt,
+) (*proto.MFAAuthenticateResponse, string, error) {
+	switch {
+	case origin == "":
+		return nil, "", trace.BadParameter("origin required")
+	case assertion == nil:
+		return nil, "", trace.BadParameter("assertion required")
+	case prompt == nil:
+		return nil, "", trace.BadParameter("prompt required")
+	case len(assertion.Response.Challenge) == 0:
+		return nil, "", trace.BadParameter("assertion challenge required")
+	case assertion.Response.RelyingPartyID == "":
+		return nil, "", trace.BadParameter("assertion relying party ID required")
+	}
+
+	allowedCreds := assertion.Response.GetAllowedCredentialIDs()
+	uv := assertion.Response.UserVerification == protocol.VerificationRequired
+	passwordless := len(allowedCreds) == 0 && uv
+
+	// Prepare challenge data for the device.
+	ccdJSON, err := json.Marshal(&CollectedClientData{
+		Type:      string(protocol.AssertCeremony),
+		Challenge: base64.RawURLEncoding.EncodeToString(assertion.Response.Challenge),
+		Origin:    origin,
+	})
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+	ccdHash := sha256.Sum256(ccdJSON)
+
+	rpID := assertion.Response.RelyingPartyID
+	var appID string
+	if val, ok := assertion.Response.Extensions[wanlib.AppIDExtension]; ok {
+		appID = fmt.Sprint(val)
+	}
+
+	// mu guards the variables below it.
+	var mu sync.Mutex
+	var assertionResp *libfido2.Assertion
+	var credentialID []byte
+	var userID []byte
+	var username string
+	var usedAppID bool
+
+	if err := runOnFIDO2Devices(
+		ctx, prompt, passwordless, /* skipAdditionalPrompt */
+		/* filter */ func(dev FIDODevice, info *deviceInfo) (bool, error) {
+			switch {
+			case uv && !info.uvCapable():
+				return false, nil
+			case passwordless && !info.rk:
+				return false, nil
+			case len(allowedCreds) == 0: // Nothing else to check
+				return true, nil
+			}
+
+			// Does the device have a suitable credential?
+			if _, err := dev.Assertion(rpID, ccdHash[:], allowedCreds, "" /* pin */, &libfido2.AssertionOpts{
```

Comment:
I see the reason for them, but I've never been a big fan of these comments annotating the parameters.

Can we rewrite this in such a way that they are not necessary?

```
pin := "" // no PIN required
if _, err := dev.Assertion(rpID, hash, allowedCreds, pin, ...)
```


Same thing for these callbacks.

```
deviceFilter := func(dev FIDODevice, info *deviceInfo) (bool, error) { ... }

if err := runOnFIDO2Devices(cts, prompt, passwordless, deviceFilter, ...)
```

Reply:
Oh, my poor inline comments, I like them so much.

In a more serious note, I have a bunch of those all over the place, so I can retract here but there will be some left. There are also a few cases where the variable names don't exactly line up, so the comment does clarify something. Give me a nudge if you feel strongly about this.

Reply:
Replaced a few where I could use consts or formatting was wonky, but kept others that I think are looking OK (or that would change the flow of the code too much).